File: Keepr.client/src/services/KeepsService.js
import { Keep } from "../models/Keep.js";
import { api } from "./AxiosService.js";
import { AppState } from '../AppState.js'
import { Profile } from "../models/Account.js";


class KeepsService {

  async GetKeeps() {
    const res = await api.get('api/keeps')
    // console.log(res.data)
    AppState.keeps = res.data.map(k => new Keep(k))
  }

  async getKeepById(keepId) {
    const res = await api.get(`api/keeps/${keepId}`)
    AppState.activeKeep = new Keep(res.data)
    console.log(AppState.activeKeep)
    return AppState.activeKeep
  }

  async deleteKeep(keepId) {
    const res = await api.delete(`api/keeps/${keepId}`)
    AppState.activeKeep = null;
    AppState.accountKeeps = AppState.accountKeeps.filter(k => k.id != keepId)
    AppState.keeps = AppState.keeps.filter(k => k.id != keepId)
  }

  async createKeep(data) {
    const res = await api.post(`api/keeps`, data)
    console.log(res.data)
    AppState.accountKeeps = [new Keep(res.data), ...AppState.accountKeeps]
    AppState.keeps = [new Keep(res.data), ...AppState.keeps]
  }

}

export const keepsService = new KeepsService();

File: Keepr/Services/KeepsService.cs
namespace Keepr.Services;

public class KeepsService
{

  private readonly KeepsRepository _keepsRepo;

  public KeepsService(KeepsRepository keepsRepo)
  {
    _keepsRepo = keepsRepo;
  }

  public Keep CreateKeep(Keep data, Account userInfo)
  {
    data.CreatorId = userInfo.Id;

    Keep keep = _keepsRepo.CreateKeep(data);
    keep.Creator = userInfo;
    return keep;
  }

  public List<Keep> GetAllKeeps()
  {
    return _keepsRepo.GetAllKeeps();
  }

  public Keep GetKeepById(int keepId, bool viewed)
  {
    Keep keep = _keepsRepo.GetKeepById(keepId);
    if (keep == null)
    {
      throw new Exception("Bad Keep Id!");
    }
    if (viewed)
    {
      keep.Views++;
      _keepsRepo.AddView(keep);
    }
    return keep;
  }

  public Keep EditKeep(int id, Keep data, string userId)
  {
    Keep keep = GetKeepById(id, false);

    if (keep.CreatorId != userId)
    {
      throw new Exception("This is not your Keep to Edit!");
    }

    keep.Name = data.Name;
    keep.Description = data.Description;
    return _keepsRepo.EditKeep(keep);
  }

  public void DeleteKeep(int id, string userId)
  {
    Keep keep = GetKeepById(id, false);
    if (keep.CreatorId != userId)
    {
      throw new Exception("This is not your Keep to Delete!");
    }

    _keepsRepo.DeleteKeep(id);
  }
}